File: swift/Packages/CanopyWorkingTree/Sources/CanopyWorkingTree/IgnorePolicy.swift
import Foundation

/// Filesystem membership for a folder: which local paths belong to its tree.
///
/// A port of `packages/fs/src/ignore-policy.ts`, which Arbor Sync applies to
/// every placed folder; `tests/fixtures/ignore-policy/cases.json` holds both
/// to the same answers. Mandatory exclusions are tooling and platform state
/// that is never tree content. User rules come from `.arborignore` and
/// `.gitignore` in the directory they apply from, in Git's pattern grammar;
/// Git itself, `.git/info/exclude` and machine-global ignore files are never
/// read. Rule files are read once, on first use beneath their directory.
/// Paths are tree paths (`/a/b`) relative to `root`.
public final class IgnorePolicy {
    public enum Membership: String, Sendable {
        case included, mandatory, ignored
    }

    public struct Decision: Equatable, Sendable {
        public var membership: Membership
        /// For `ignored`: the tree path of the ignore file whose rule decided.
        public var source: String?
        /// For `ignored`: that rule as written.
        public var pattern: String?
    }

    /// An ignore file whose rules do not apply. It names the file, never its contents.
    public struct Diagnostic: Equatable, Sendable {
        public var code: String
        public var path: String
        public var message: String
    }

    /// Directory names that are never tree content, wherever they appear.
    public static let mandatoryDirectoryNames: Set<String> = [".git", "node_modules", ".arbor", "Trash", ".build", "DerivedData"]
    /// Ignore-rule files, lowest precedence first: `.arborignore` wins over `.gitignore` beside it.
    public static let ignoreFileNames = [".gitignore", ".arborignore"]

    public let root: URL
    public private(set) var diagnostics: [Diagnostic] = []
    private let excluded: [String]
    private var files: [String: [RuleFile]] = [:]
    private var directories: [String: Decision] = [:]

    /// `excludedRoots` are nested tree mounts, whose content belongs to another tree.
    public init(root: URL, excludedRoots: [URL] = []) {
        let root = root.standardizedFileURL.resolvingSymlinksInPath()
        self.root = root
        excluded = excludedRoots.compactMap { url in
            let path = url.standardizedFileURL.resolvingSymlinksInPath().path
            guard path.hasPrefix(root.path + "/") else { return nil }
            return String(path.dropFirst(root.path.count))
        }
    }

    /// Whether `treePath` is tree content, never tree content, or kept out by a user rule.
    public func decision(_ treePath: String, isDirectory: Bool) -> Decision {
        let segments = Self.segments(treePath)
        if segments.isEmpty { return .included }
        if mandatory(segments, isDirectory: isDirectory) { return .mandatory }
        return ruled(segments, isDirectory: isDirectory)
    }

    static func isTransactionTemporaryName(_ name: String) -> Bool {
        name.contains(".arbor-write-") || name.contains(".arbor-txn-")
    }

    static func isCloudPlaceholderName(_ name: String) -> Bool {
        name.hasPrefix(".") && name.hasSuffix(".icloud") && name.count > ".icloud".count + 1
    }

    private func mandatory(_ segments: [String], isDirectory: Bool) -> Bool {
        let path = Self.treePath(segments)
        if excluded.contains(where: { path == $0 || path.hasPrefix($0 + "/") }) { return true }
        return segments.indices.contains { index in
            let name = segments[index]
            return Self.isTransactionTemporaryName(name)
                || Self.isCloudPlaceholderName(name)
                || ((index < segments.count - 1 || isDirectory) && Self.mandatoryDirectoryNames.contains(name))
        }
    }

    private func ruled(_ segments: [String], isDirectory: Bool) -> Decision {
        if segments.count > 1 {
            let parent = directory(Array(segments.dropLast()))
            if parent.membership != .included { return parent }
        }
        if !isDirectory, Self.ignoreFileNames.contains(segments[segments.count - 1]) { return .included }
        // Deeper files override shallower ones, and within a file the last matching rule wins.
        for depth in stride(from: segments.count - 1, through: 0, by: -1) {
            let relativePath = segments[depth...].joined(separator: "/")
            for file in rules(in: Array(segments[..<depth])).reversed() {
                for rule in file.rules.reversed() where rule.matches(relativePath, isDirectory: isDirectory) {
                    return rule.negative ? .included : Decision(membership: .ignored, source: file.source, pattern: rule.pattern)
                }
            }
        }
        return .included
    }

    private func directory(_ segments: [String]) -> Decision {
        let key = Self.treePath(segments)
        if let decision = directories[key] { return decision }
        let decision = ruled(segments, isDirectory: true)
        directories[key] = decision
        return decision
    }

    private func rules(in segments: [String]) -> [RuleFile] {
        let key = Self.treePath(segments)
        if let loaded = files[key] { return loaded }
        let loaded = Self.ignoreFileNames.compactMap { readRules(segments + [$0]) }
        files[key] = loaded
        return loaded
    }

    private func readRules(_ segments: [String]) -> RuleFile? {
        let source = Self.treePath(segments)
        let url = segments.reduce(root) { $0.appendingPathComponent($1) }
        // A symbolic link is not tree content, so its target contributes no rules.
        guard let attributes = try? FileManager.default.attributesOfItem(atPath: url.path),
              attributes[.type] as? FileAttributeType == .typeRegular else { return nil }
        guard let data = try? Data(contentsOf: url) else {
            diagnostics.append(Diagnostic(code: "ignore-file-unreadable", path: source,
                                          message: "\(source) could not be read, so none of its ignore rules apply."))
            return nil
        }
        guard let text = String(data: data, encoding: .utf8) else {
            diagnostics.append(Diagnostic(code: "ignore-file-not-utf8", path: source,
                                          message: "\(source) is not valid UTF-8, so none of its ignore rules apply."))
            return nil
        }
        return RuleFile(source: source, rules: Rule.parse(text))
    }

    private static func segments(_ treePath: String) -> [String] {
        treePath.split(separator: "/", omittingEmptySubsequences: true).map(String.init)
    }

    private static func treePath(_ segments: [String]) -> String {
        "/" + segments.joined(separator: "/")
    }
}

extension IgnorePolicy.Decision {
    static let included = IgnorePolicy.Decision(membership: .included)
    static let mandatory = IgnorePolicy.Decision(membership: .mandatory)
}

private struct RuleFile {
    var source: String
    var rules: [Rule]
}

/// One line of an ignore file, compiled to a regular expression with Git's
/// wildmatch semantics (`WM_PATHNAME`). The expression uses only syntax that
/// ICU and JavaScript read alike, so it matches what the TypeScript rule does.
private struct Rule {
    var pattern: String
    var negative: Bool
    var directoryOnly: Bool
    /// Match the basename at any depth, rather than the path from the rule's directory.
    var basename: Bool
    var regex: NSRegularExpression

    func matches(_ relativePath: String, isDirectory: Bool) -> Bool {
        if directoryOnly && !isDirectory { return false }
        let subject = basename ? String(relativePath.split(separator: "/", omittingEmptySubsequences: false).last ?? "") : relativePath
        let range = NSRange(subject.startIndex..<subject.endIndex, in: subject)
        return regex.firstMatch(in: subject, options: [.anchored], range: range) != nil
    }

    /// One ignore file's rules, in file order.
    static func parse(_ text: String) -> [Rule] {
        var text = text
        if text.unicodeScalars.first == "\u{FEFF}" { text.unicodeScalars.removeFirst() }
        var rules: [Rule] = []
        // By scalar: "\r\n" is one Character, so a Character split would miss it.
        for raw in text.unicodeScalars.split(separator: "\n", omittingEmptySubsequences: false) {
            var scalars = Array(raw)
            if scalars.last == "\r" { scalars.removeLast() }
            scalars = trimTrailingSpaces(scalars)
            if scalars.isEmpty || scalars[0] == "#" { continue }
            let line = string(scalars)
            var body = scalars
            let negative = body.first == "!"
            if negative { body.removeFirst() }
            let directoryOnly = body.last == "/"
            if directoryOnly { body.removeLast() }
            if body.isEmpty { continue }
            let basename = !body.contains("/")
            if body.first == "/" { body.removeFirst() }
            // `\z` ends the input exactly, as `$` does in JavaScript without the `m` flag.
            guard let source = wildmatchSource(body),
                  let regex = try? NSRegularExpression(pattern: "^(?:\(source))\\z") else { continue }
            rules.append(Rule(pattern: line, negative: negative, directoryOnly: directoryOnly, basename: basename, regex: regex))
        }
        return rules
    }

    /// Remove trailing spaces that are not escaped with a backslash.
    private static func trimTrailingSpaces(_ line: [Unicode.Scalar]) -> [Unicode.Scalar] {
        var end = 0
        var index = 0
        while index < line.count {
            if line[index] == "\\" {
                index += 1
                end = min(index + 1, line.count)
            } else if line[index] != " " {
                end = index + 1
            }
            index += 1
        }
        return Array(line[..<end])
    }

    private static func string(_ scalars: some Sequence<Unicode.Scalar>) -> String {
        var result = ""
        result.unicodeScalars.append(contentsOf: scalars)
        return result
    }

    /// A regular-expression literal for one scalar. A backslash before ASCII
    /// punctuation is always literal in ICU, inside a set too, where `&&`,
    /// `--` and `{}` would otherwise be set syntax.
    private static func escape(_ scalar: Unicode.Scalar) -> String {
        let punctuation = (0x21...0x7E).contains(scalar.value) && !scalar.properties.isAlphabetic && !("0"..."9").contains(scalar)
        return punctuation ? "\\" + String(scalar) : String(scalar)
    }

    private static let posixClasses: [String: String] = [
        "alnum": "a-zA-Z0-9",
        "alpha": "a-zA-Z",
        "blank": " \\t",
        "digit": "0-9",
        "lower": "a-z",
        "space": " \\t\\n\\r\\f\\x0B",
        "upper": "A-Z",
        "xdigit": "0-9a-fA-F",
    ]

    /// Git's wildmatch as a regular-expression body; nil when the pattern can match nothing.
    private static func wildmatchSource(_ characters: [Unicode.Scalar]) -> String? {
        var source = ""
        var index = 0
        while index < characters.count {
            let character = characters[index]
            if character == "*" {
                var end = index
                while end < characters.count && characters[end] == "*" { end += 1 }
                let doubled = end - index >= 2
                let opensSegment = index == 0 || characters[index - 1] == "/"
                let closesSegment = end == characters.count || characters[end] == "/"
                if doubled && opensSegment && closesSegment {
                    if end == characters.count {
                        source += "[\\s\\S]*"
                        index = end
                    } else {
                        source += "(?:[\\s\\S]*/)?"
                        index = end + 1
                    }
                    continue
                }
                source += "[^/]*"
                index = end
                continue
            }
            if character == "?" {
                source += "[^/]"
                index += 1
                continue
            }
            if character == "[" {
                guard let (fragment, next) = bracket(characters, start: index) else { return nil }
                source += fragment
                index = next
                continue
            }
            if character == "\\" {
                guard index + 1 < characters.count else { return nil }
                source += escape(characters[index + 1])
                index += 2
                continue
            }
            source += escape(character)
            index += 1
        }
        return source
    }

    /// A bracket expression beginning at `start` (the `[`) and the index after
    /// it; nil when it is malformed or has no members, which makes the whole
    /// pattern match nothing, as in Git.
    private static func bracket(_ characters: [Unicode.Scalar], start: Int) -> (String, Int)? {
        func at(_ index: Int) -> Unicode.Scalar? { index < characters.count ? characters[index] : nil }
        var index = start + 1
        let negated = at(index) == "!" || at(index) == "^"
        if negated { index += 1 }
        var members = ""
        var first = true
        while index < characters.count {
            var character = characters[index]
            if character == "]" && !first {
                // A bracket expression never matches `/`.
                if !negated && members.isEmpty { return nil }
                return (negated ? "[^/\(members)]" : "(?:(?!/)[\(members)])", index + 1)
            }
            first = false
            if character == "[" && at(index + 1) == ":" {
                guard let close = characters[(index + 2)...].firstIndex(of: ":"), at(close + 1) == "]",
                      let range = posixClasses[string(characters[(index + 2)..<close])] else { return nil }
                members += range
                index = close + 2
                continue
            }
            if character == "\\" {
                index += 1
                guard index < characters.count else { return nil }
                character = characters[index]
            }
            if at(index + 1) == "-", let rangeEnd = at(index + 2), rangeEnd != "]" {
                var end = rangeEnd
                var after = index + 3
                if end == "\\" {
                    guard let escaped = at(index + 3) else { return nil }
                    end = escaped
                    after = index + 4
                }
                // An inverted range matches nothing.
                if character.value <= end.value { members += escape(character) + "-" + escape(end) }
                index = after
                continue
            }
            members += escape(character)
            index += 1
        }
        return nil
    }
}

File: packages/fs/src/ignore-policy.ts
import { lstat, readFile, realpath } from "node:fs/promises";
import { join, relative, resolve, sep } from "node:path";
import { decodeProtocolDirectory, type Diagnostic, type ObjectHash, type ProtocolDirectoryEntry } from "@overstory/protocol";

/**
 * Filesystem membership: which local paths belong to a tree. Discovery,
 * listing, watching, snapshots, object reads and materialization all ask one
 * policy, so a path is either tree content everywhere or nowhere.
 *
 * Mandatory exclusions are tooling and platform state that is never tree
 * content. User rules come from `.arborignore` (Overstory's portable
 * spelling) and `.gitignore` (read for compatibility) in the directory they
 * apply from; both use Git's pattern grammar. Git itself, its index,
 * `.git/info/exclude` and machine-global ignore files are never consulted, so
 * the same folder yields the same tree on every device.
 */

/** Directory names that are never tree content, wherever they appear. */
export const MANDATORY_DIRECTORY_NAMES: ReadonlySet<string> = new Set([
  ".git",
  "node_modules",
  ".arbor",
  "Trash",
  ".build",
  "DerivedData",
]);

/** Ignore-rule files, lowest precedence first: `.arborignore` wins over `.gitignore` beside it. */
export const IGNORE_FILE_NAMES = [".gitignore", ".arborignore"] as const;

export function isIgnoreFileName(name: string): boolean {
  return (IGNORE_FILE_NAMES as readonly string[]).includes(name);
}

/** An Overstory write or transaction temporary. */
export function isTransactionTemporaryName(name: string): boolean {
  return name.includes(".arbor-write-") || name.includes(".arbor-txn-");
}

/**
 * macOS Finder and filesystem metadata: `.DS_Store`, and the AppleDouble
 * `._name` companions written on volumes without extended attributes. The
 * system rewrites them on its own, so they are never tree content.
 */
export function isPlatformMetadataName(name: string): boolean {
  return name === ".DS_Store" || (name.startsWith("._") && name.length > 2);
}

/** iCloud's stand-in `.name.icloud` for an evicted file. */
export function isCloudPlaceholderName(name: string): boolean {
  return name.startsWith(".") && name.endsWith(".icloud") && name.length > ".icloud".length + 1;
}

export type Membership = "included" | "mandatory" | "ignored";

export interface IgnoreDecision {
  membership: Membership;
  /** For `ignored`: the tree path of the ignore file whose rule decided. */
  source?: string;
  /** For `ignored`: that rule as written. */
  pattern?: string;
}

interface Rule {
  pattern: string;
  negative: boolean;
  directoryOnly: boolean;
  /** Match the basename at any depth, rather than the path from the rule's directory. */
  basename: boolean;
  regex: RegExp;
}

interface RuleFile { source: string; rules: Rule[] }
interface ChainedRules { depth: number; file: RuleFile }

const INCLUDED: IgnoreDecision = { membership: "included" };
const MANDATORY: IgnoreDecision = { membership: "mandatory" };

function escapeLiteral(character: string): string {
  return /[\^$\\.*+?()[\]{}|/]/.test(character) ? `\\${character}` : character;
}

function escapeClassMember(character: string): string {
  return /[\\\]\[^\-]/.test(character) ? `\\${character}` : character;
}

const POSIX_CLASSES: Record<string, string> = {
  alnum: "a-zA-Z0-9",
  alpha: "a-zA-Z",
  blank: " \\t",
  digit: "0-9",
  lower: "a-z",
  space: " \\t\\n\\r\\f\\x0B",
  upper: "A-Z",
  xdigit: "0-9a-fA-F",
};

/**
 * A bracket expression beginning at `characters[start]` (the `[`), as a
 * regular-expression fragment and the index after it. Null when the
 * expression is malformed, which makes the whole pattern match nothing, as in
 * Git.
 */
function bracket(characters: readonly string[], start: number): { source: string; next: number } | null {
  let index = start + 1;
  const negated = characters[index] === "!" || characters[index] === "^";
  if (negated) index++;
  let members = "";
  let first = true;
  while (index < characters.length) {
    let character = characters[index]!;
    if (character === "]" && !first) {
      // A bracket expression never matches `/`; one with no members matches nothing.
      if (!negated && !members) return null;
      return { source: negated ? `[^/${members}]` : `(?:(?!/)[${members}])`, next: index + 1 };
    }
    first = false;
    if (character === "[" && characters[index + 1] === ":") {
      const close = characters.indexOf(":", index + 2);
      if (close < 0 || characters[close + 1] !== "]") return null;
      const range = POSIX_CLASSES[characters.slice(index + 2, close).join("")];
      if (range === undefined) return null;
      members += range;
      index = close + 2;
      continue;
    }
    if (character === "\\") {
      index++;
      if (index >= characters.length) return null;
      character = characters[index]!;
    }
    if (characters[index + 1] === "-" && characters[index + 2] !== undefined && characters[index + 2] !== "]") {
      let end = characters[index + 2]!;
      let after = index + 3;
      if (end === "\\") {
        end = characters[index + 3] ?? "";
        after = index + 4;
        if (!end) return null;
      }
      // An inverted range matches nothing.
      if (character.codePointAt(0)! <= end.codePointAt(0)!) members += `${escapeClassMember(character)}-${escapeClassMember(end)}`;
      index = after;
      continue;
    }
    members += escapeClassMember(character);
    index++;
  }
  return null;
}

/** Git's wildmatch with `WM_PATHNAME`, as a regular-expression body; null when the pattern can match nothing. */
function wildmatchSource(glob: string): string | null {
  const characters = Array.from(glob);
  let source = "";
  let index = 0;
  while (index < characters.length) {
    const character = characters[index]!;
    if (character === "*") {
      let end = index;
      while (characters[end] === "*") end++;
      const doubled = end - index >= 2;
      const opensSegment = index === 0 || characters[index - 1] === "/";
      const closesSegment = end === characters.length || characters[end] === "/";
      if (doubled && opensSegment && closesSegment) {
        if (end === characters.length) { source += "[\\s\\S]*"; index = end; continue; }
        source += "(?:[\\s\\S]*/)?";
        index = end + 1;
        continue;
      }
      source += "[^/]*";
      index = end;
      continue;
    }
    if (character === "?") { source += "[^/]"; index++; continue; }
    if (character === "[") {
      const parsed = bracket(characters, index);
      if (!parsed) return null;
      source += parsed.source;
      index = parsed.next;
      continue;
    }
    if (character === "\\") {
      if (index + 1 >= characters.length) return null;
      source += escapeLiteral(characters[index + 1]!);
      index += 2;
      continue;
    }
    source += escapeLiteral(character);
    index++;
  }
  return source;
}

/** Remove trailing spaces that are not escaped with a backslash. */
function trimTrailingSpaces(line: string): string {
  let end = 0;
  for (let index = 0; index < line.length; index++) {
    if (line[index] === "\\") { index++; end = Math.min(index + 1, line.length); continue; }
    if (line[index] !== " ") end = index + 1;
  }
  return line.slice(0, end);
}

/** One ignore file's rules, in file order. */
function parseIgnoreRules(text: string): Rule[] {
  const rules: Rule[] = [];
  for (const raw of text.replace(/^\uFEFF/, "").split("\n")) {
    const line = trimTrailingSpaces(raw.endsWith("\r") ? raw.slice(0, -1) : raw);
    if (!line || line.startsWith("#")) continue;
    let body = line;
    const negative = body.startsWith("!");
    if (negative) body = body.slice(1);
    const directoryOnly = body.endsWith("/");
    if (directoryOnly) body = body.slice(0, -1);
    if (!body) continue;
    const basename = !body.includes("/");
    if (body.startsWith("/")) body = body.slice(1);
    const source = wildmatchSource(body);
    if (source === null) continue;
    rules.push({ pattern: line, negative, directoryOnly, basename, regex: new RegExp(`^${source}$`, "u") });
  }
  return rules;
}

function ruleMatches(rule: Rule, relativePath: string, isDirectory: boolean): boolean {
  if (rule.directoryOnly && !isDirectory) return false;
  if (rule.basename) return rule.regex.test(relativePath.slice(relativePath.lastIndexOf("/") + 1));
  return rule.regex.test(relativePath);
}

function segmentsOf(treePath: string): string[] {
  return treePath.split("/").filter(Boolean);
}

function treePathOf(segments: readonly string[]): string {
  return `/${segments.join("/")}`;
}

/**
 * One operation's view of a folder's membership. Rule files are read once, on
 * first use beneath their directory, and never re-read: load a new policy to
 * see edits. Paths are tree paths (`/a/b`) relative to `root`.
 */
export class IgnorePolicy {
  readonly diagnostics: Diagnostic[] = [];
  private readonly files = new Map<string, Promise<RuleFile[]>>();
  private readonly chains = new Map<string, Promise<ChainedRules[]>>();
  private readonly directories = new Map<string, Promise<IgnoreDecision>>();

  constructor(
    readonly root: string,
    private readonly excluded: readonly string[],
    private readonly read?: IgnoreFileReader,
  ) {}

  /** Whether `treePath` is tree content, never tree content, or kept out by a user rule. */
  async decision(treePath: string, isDirectory: boolean): Promise<IgnoreDecision> {
    const segments = segmentsOf(treePath);
    if (!segments.length) return INCLUDED;
    if (this.mandatory(segments, isDirectory)) return MANDATORY;
    return this.ruled(segments, isDirectory);
  }

  private mandatory(segments: readonly string[], isDirectory: boolean): boolean {
    const path = treePathOf(segments);
    if (this.excluded.some((excluded) => path === excluded || path.startsWith(`${excluded}/`))) return true;
    return segments.some((name, index) =>
      isTransactionTemporaryName(name)
      || isCloudPlaceholderName(name)
      || isPlatformMetadataName(name)
      || ((index < segments.length - 1 || isDirectory) && MANDATORY_DIRECTORY_NAMES.has(name))
    );
  }

  private async ruled(segments: readonly string[], isDirectory: boolean): Promise<IgnoreDecision> {
    if (segments.length > 1) {
      const parent = await this.directory(segments.slice(0, -1));
      if (parent.membership !== "included") return parent;
    }
    if (!isDirectory && isIgnoreFileName(segments.at(-1)!)) return INCLUDED;
    // Deeper files override shallower ones, and within a file the last matching rule wins.
    const chain = await this.chain(segments.slice(0, -1));
    for (let index = chain.length - 1; index >= 0; index--) {
      const { depth, file } = chain[index]!;
      const relativePath = segments.slice(depth).join("/");
      for (let ruleIndex = file.rules.length - 1; ruleIndex >= 0; ruleIndex--) {
        const rule = file.rules[ruleIndex]!;
        if (!ruleMatches(rule, relativePath, isDirectory)) continue;
        return rule.negative ? INCLUDED : { membership: "ignored", source: file.source, pattern: rule.pattern };
      }
    }
    return INCLUDED;
  }

  /** The rule files that apply within a directory, shallowest first, each with the depth it applies from. */
  private chain(segments: readonly string[]): Promise<ChainedRules[]> {
    const key = treePathOf(segments);
    let chain = this.chains.get(key);
    if (!chain) {
      const parent = segments.length ? this.chain(segments.slice(0, -1)) : Promise.resolve([]);
      chain = Promise.all([parent, this.rulesIn(segments)]).then(([above, own]) =>
        own.length ? [...above, ...own.map((file) => ({ depth: segments.length, file }))] : above
      );
      this.chains.set(key, chain);
    }
    return chain;
  }

  private directory(segments: readonly string[]): Promise<IgnoreDecision> {
    const key = treePathOf(segments);
    let decision = this.directories.get(key);
    if (!decision) {
      decision = this.ruled(segments, true);
      this.directories.set(key, decision);
    }
    return decision;
  }

  private rulesIn(segments: readonly string[]): Promise<RuleFile[]> {
    const key = treePathOf(segments);
    let files = this.files.get(key);
    if (!files) {
      files = Promise.all(IGNORE_FILE_NAMES.map((name) => this.readRules([...segments, name])))
        .then((loaded) => loaded.filter((file): file is RuleFile => file !== null));
      this.files.set(key, files);
    }
    return files;
  }

  private async readRules(segments: readonly string[]): Promise<RuleFile | null> {
    const source = treePathOf(segments);
    const absolute = join(this.root, ...segments);
    let bytes: Uint8Array;
    try {
      if (this.read) {
        const read = await this.read(source);
        if (!read) return null;
        bytes = read;
      } else {
        // A symbolic link is not tree content, so its target contributes no rules.
        if (!(await lstat(absolute)).isFile()) return null;
        bytes = await readFile(absolute);
      }
    } catch (error) {
      const code = (error as NodeJS.ErrnoException).code;
      if (code === "ENOENT" || code === "ENOTDIR") return null;
      this.diagnostics.push({
        code: "ignore-file-unreadable",
        path: source,
        severity: "warning",
        message: `${source} could not be read, so none of its ignore rules apply.`,
      });
      return null;
    }
    let text: string;
    try {
      text = new TextDecoder("utf-8", { fatal: true, ignoreBOM: true }).decode(bytes);
    } catch {
      this.diagnostics.push({
        code: "ignore-file-not-utf8",
        path: source,
        severity: "warning",
        message: `${source} is not valid UTF-8, so none of its ignore rules apply.`,
      });
      return null;
    }
    return { source, rules: parseIgnoreRules(text) };
  }
}

/** An ignore file's bytes by tree path, or null when there is none. */
export type IgnoreFileReader = (treePath: string) => Promise<Uint8Array | null>;

/**
 * Load a folder's membership policy for one operation. `excludedRoots` are
 * nested tree mounts: absolute paths whose content belongs to another tree.
 * `read` takes ignore files from elsewhere than the folder, such as a root
 * about to be written to it.
 */
export async function loadIgnorePolicy(
  path: string,
  options: { excludedRoots?: readonly string[]; read?: IgnoreFileReader } = {},
): Promise<IgnorePolicy> {
  const root = await realpath(path);
  const excluded: string[] = [];
  for (const item of options.excludedRoots ?? []) {
    const absolute = await realpath(item).catch(() => resolve(item));
    const remainder = relative(root, absolute);
    if (!remainder || remainder === ".." || remainder.startsWith(`..${sep}`)) continue;
    excluded.push(`/${remainder.split(sep).join("/")}`);
  }
  return new IgnorePolicy(root, excluded, options.read);
}

/**
 * Whether a walk leaves out a path, by tree path relative to the walked root.
 * `snapshotDirectory` and `materializeTree` call it only for ordinary entries
 * they would otherwise include or delete.
 */
export type SkipPath = (treePath: string, isDirectory: boolean) => Promise<boolean>;

/** Whether a path of that kind is in the root a folder last held; tracked paths stay synchronized when a rule matches them. */
export type TrackedLookup = (treePath: string, isDirectory: boolean) => Promise<boolean>;

/**
 * The walk filter for a policy: an ignored path is left out unless it is
 * tracked. Tracked is asked only for ignored paths, so an ordinary walk loads
 * nothing extra. `prefix` is the walked root's tree path under the policy's
 * root.
 */
export function membershipSkip(policy: IgnorePolicy, tracked: TrackedLookup | null, prefix = "/"): SkipPath {
  const base = segmentsOf(prefix);
  return async (treePath, isDirectory) => {
    const path = treePathOf([...base, ...segmentsOf(treePath)]);
    const { membership } = await policy.decision(path, isDirectory);
    if (membership === "included") return false;
    if (membership === "mandatory") return true;
    return !(tracked && await tracked(path, isDirectory));
  };
}

/**
 * Tracked paths of `root`: an entry of the same kind at the path. A file that
 * became a directory (or the reverse) is a new, untracked entry. Directory
 * objects are loaded on demand and remembered for this lookup's lifetime.
 */
export function trackedEntries(root: ObjectHash, load: (hash: ObjectHash) => Promise<Uint8Array>): TrackedLookup {
  const entryAt = entries(root, load);
  return async (treePath, isDirectory) => {
    const entry = await entryAt(treePath);
    return isDirectory ? entry?.directory !== undefined : entry?.file !== undefined;
  };
}

/** Ignore files as a root holds them, for the policy a folder will have once that root is written. */
export function ignoreFilesIn(root: ObjectHash, load: (hash: ObjectHash) => Promise<Uint8Array>): IgnoreFileReader {
  const entryAt = entries(root, load);
  return async (treePath) => {
    const entry = await entryAt(treePath);
    return entry?.file === undefined ? null : load(entry.file as ObjectHash);
  };
}

/** The entry at a tree path beneath `root`, loading and remembering directories on demand. */
function entries(root: ObjectHash, load: (hash: ObjectHash) => Promise<Uint8Array>): (treePath: string) => Promise<ProtocolDirectoryEntry | undefined> {
  const directories = new Map<ObjectHash, Promise<Map<string, ProtocolDirectoryEntry>>>();
  const children = (hash: ObjectHash) => {
    let entries = directories.get(hash);
    if (!entries) {
      entries = load(hash).then((bytes) => {
        const object = decodeProtocolDirectory(bytes);
        if (object.type !== "directory") throw new Error(`Expected a directory: ${hash}`);
        return new Map(object.entries.map((entry) => [entry.name, entry]));
      });
      entries.catch(() => directories.delete(hash));
      directories.set(hash, entries);
    }
    return entries;
  };
  return async (treePath) => {
    let entry: ProtocolDirectoryEntry | undefined = { name: "", directory: root };
    for (const name of segmentsOf(treePath)) {
      if (!entry?.directory) return undefined;
      entry = (await children(entry.directory as ObjectHash)).get(name);
    }
    return entry;
  };
}
